src/documentation/programming-guide.md

```diff
@@ -532,35 +533,173 @@ tree, [2]
 
 Thus, `GroupByKey` represents a transform from a multimap (multiple keys to individual values) to a uni-map (unique keys to collections of values).
 
-##### **Joins with CoGroupByKey**
+#### <a name="transforms-cgbk"></a> **Joins with CoGroupByKey**
+
+`CoGroupByKey` performs a relational join of two or more key/value
+`PCollection`s that have the same key type.
+[Design Your Pipeline]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources)
+shows an example pipeline that uses a join.
+
+Consider using `CoGroupByKey` if you have multiple data sets that provide
+information about related things. For example, let's say you have two different
+files with auction data: one file has the auction ID, bid data, and pricing
+data; the other file has the auction ID and an item description. You can join
+those two data sets, using the auction ID as a common key and the other data as
+the associated values. After the join, you have one data set that contains all
+of the information (bid, price, and item) associated with each auction ID.
+
+<span class="language-java">
+In the Beam SDK for Java, `CoGroupByKey` accepts a tuple of keyed
+`PCollection`s (`PCollection<KV<K, V>>`) as input. For type safety, the SDK
+requires you to pass each `PCollection` as part of a `KeyedPCollectionTuple`.
+You must declare a `TupleTag` for each input `PCollection` in the
+`KeyedPCollectionTuple` that you want to pass to `CoGroupByKey`.
+</span>
+<span class="language-java">
+As output, `CoGroupByKey` returns a `PCollection<KV<K, CoGbkResult>>`, which
+groups values from all the input `PCollection`s by their common keys. Each key
+(all of type `K`) will have a different `CoGbkResult`, which is a map from
+`TupleTag<T>` to `Iterable<T>`. You can access a specific collection in an
+`CoGbkResult` object by using the `TupleTag` that you supplied with the initial
+collection.
+</span>
+
+<span class="language-py">
+In the Beam SDK for Python, `CoGroupByKey` accepts a dictionary of keyed
+`PCollection`s as input. As output, `CoGroupByKey` creates a single output
+`PCollection` that contains one key/value tuple for each key in the input
+`PCollection`s. Each key's value is a dictionary that maps each tag to an
+iterable of the values under they key in the corresponding `PCollection`.
+</span>
+
+The following example uses two input collections to show the mechanics of
+`CoGroupByKey` and tags.
+
+<span class="language-java">
+The first input collection `PCollection<KV<K, V1>>` has a `TupleTag<V1>` called
+`tag1` and contains the following key/value pairs:
+</span>
+<span class="language-py">
+The first input `PCollection` has tag `tag1` and contains the following
+key/value pairs:
+</span>
+```
+   User1, Address1
```

Comment:
How about we have a common example for this explanation as well as the code samples?

e.g.
```
emails
  amy, amy@example.com
  carl, carl@example.com
  julia, julia@example.com
  carl, carl@email.com
```

```
phones
  amy, 111-222-3333
  james, 222-333-4444
  amy, 333-444-5555
  carl, 444-555-6666
```

```
CoGroupByKey
  amy: {
    emails: [amy@example.com]
    phones: [111-222-3333, 333-444-5555]
  }
  carl: {
    emails: [carl@example.com, carl@email.com]
    phones: [444-555-6666]
  }
  james: {
    emails: []
    phones: [222-333-4444]
  }
  julia: {
    emails: [julia@example.com]
    phones: []
  }
```

And then have both the Java and Python snippets actually use those values.

Reply:
Great, I'll update this and use the examples in your updated Python snippets PR so it all matches